cirq/value/linear_dict.py

```diff
@@ -0,0 +1,173 @@
+    """
+    def __init__(self, terms: Mapping[Any, Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of abstract vectors to coefficients in the linear
+                combination being initialized.
+        """
+        super(LinearDict, self).__init__()
+        self.update(terms)
+
+    def clean(self, *, tolerance: float=1e-9) -> None:
+        """Modifies self by deleting terms with small coefficients."""
+        negligible = [v for v, c in self.items() if abs(c) <= tolerance]
+        for v in negligible:
+            del self[v]
+
+    def copy(self) -> 'LinearDict':
+        return LinearDict(super(LinearDict, self).copy())
+
+    def __getitem__(self, vector: Any) -> Scalar:
+        if vector in self:
+            return super(LinearDict, self).__getitem__(vector)
+        return 0.0
+
+    def __iadd__(self, other: Any) -> 'LinearDict':
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
+
+        for vector, coefficient in other.items():
+            self[vector] = self[vector] + coefficient
+        self.clean(tolerance=0)
+        return self
+
+    def __add__(self, other: Any) -> 'LinearDict':
+        result = self.copy()
+        result += other
+        return result
+
+    def __isub__(self, other: Any) -> 'LinearDict':
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
+
+        for vector, coefficient in other.items():
+            self[vector] = self[vector] - coefficient
+        self.clean(tolerance=0)
+        return self
+
+    def __sub__(self, other: Any) -> 'LinearDict':
+        result = self.copy()
+        result -= other
+        return result
+
+    def __neg__(self) -> 'LinearDict':
+        return LinearDict({v: -c for v, c in self.items()})
+
+    def __imul__(self, a: Scalar) -> 'LinearDict':
+        for vector in self:
+            self[vector] *= a
+        self.clean(tolerance=0)
+        return self
+
+    def __mul__(self, a: Scalar) -> 'LinearDict':
+        result = self.copy()
+        result *= a
+        return result
+
+    def __rmul__(self, a: Scalar) -> 'LinearDict':
+        return self.__mul__(a)
+
+    def __truediv__(self, a: Scalar) -> 'LinearDict':
+        return self.__mul__(1 / a)
+
+    def __bool__(self) -> bool:
+        return not all(c == 0 for c in self.values())
+
+    def __eq__(self, other: Any) -> bool:
+        """Checks whether two linear combinations are exactly equal.
+
+        Presence or absence of terms with coefficients exactly equal to
+        zero does not affect outcome.
+
+        Not appropriate for most practical purposes due to sensitivity to
+        numerical error in floating point coefficients. Use cirq.approx_eq()
+        instead.
+        """
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
```

Comment:
*Definitely* don't do this. For example, this means LinearDict behaves unlike other values when nested. A singleton list is not equal to the one item in the list, and we should behave the same way.

If a value wants to compare as equal to a LinearDict that value should be the one to special case LinearDict instead of vice versa.

Reply:
Agreed.

Reply:
Done.